[feature] OnlineDB ConfigurationManager

lmoureaux opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->
Add a new class ConfigurationManager as the main entry point for gemonlinedb. Use the opportunity to better define the exception and thread safety policies within gemonlinedb.

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue(s) here. -->
Requested by Jared to implement hardware objects.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Unit tests.

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

<!--- Template thanks to https://www.talater.com/open-source-templates/#/page/99 -->

lmoureaux committed:
Make getFileInPath return absolute paths

lmoureaux committed:
Add gem::onlinedb::ConfigurationManager

lmoureaux committed:
Make configuration access thread-safe

lmoureaux committed:
Make it easier to catch Xerces exceptions

lmoureaux committed:
Use Xerces exceptions wrapper in tests

lmoureaux committed:
Simplify and document exceptions in ConfigurationManager

lmoureaux committed:
Update gemonlinedb.h

* Add some important classes that were missing
* Mention ConfigurationManager
* Document exceptions
* Document thread safety

lmoureaux committed:
Make XercesGuard thread-safe

It is being used extensively and calls unsafe static functions. See also
http://xerces.apache.org/xerces-c/faq-parse-3.html#faq-6

lmoureaux committed:
Drop unused onlinedb exceptions

mexanick commented on gemonlinedb/include/gem/onlinedb/detail/XMLUtils.h:
```diff
@@ -143,6 +148,50 @@ namespace gem {
+             *        object into @ref ParseError.
+             */
+            template<class F, class... Args>
+            auto xercesExceptionsToXcept(F f, Args... args) ->
```
Neatpicking here: avoid if possible returning `auto` types

lmoureaux replied:
Let me know if I'm wrong, but I don't think it is actually possible here.

mexanick replied:
Ah, I missed that line 164 in cursory look. Then, fine. Sometimes in the future should move to c++14 (maybe in 10 years :D)

lmoureaux replied:
Actually xDAQ 15 should bring C++14 support.

mexanick commented on gemonlinedb/include/gem/onlinedb/detail/XMLUtils.h:
```diff
@@ -143,6 +148,50 @@ namespace gem {
+             *        object into @ref std::runtime_error.
+             */
+            template<class F, class... Args>
+            auto xercesExceptionsToStd(F f, Args... args) ->
```
same as above

mexanick commented on gemonlinedb/src/common/ConfigurationManager.cc:
```diff
@@ -0,0 +1,167 @@
+                    "xml/schema:/opt/cmsgemos/share/gemonlinedb/xml/schema");
+
+                // Find the file (throws if not found)
+                auto filePath = detail::getFileInPath("system-topology.xml", path);
```
is this `std::string` or `char*`? If so, use explicit type declaration instead of `auto`

lmoureaux replied:
If it were `char*` I would have changed `getFileInPath()` to return `std::string` ;)

bdorney replied:
I think @mexanick's point was to explicitly illustrate the type and not use `auto`.

lmoureaux replied:
What would be a reason not to use `auto` as recommended by many C++ 11 tutorials?

mexanick replied:
The reason is that in this case it is not obvious what's the returned type (e.g. contrary to nearby constructions like `make_shared...`

lmoureaux replied:
You wouldn't know the actual type being returned, only that it is compatible with `std::string`. Now consider the following snippets:
#### 1. No auto
```cpp
char *foo();
std::string bar = foo();
std::cout << bar.size() << " " << (bar + "coming from foo") << std::endl;
```
#### 2. auto
```cpp
char *foo();
auto bar = foo();
std::cout << bar.size() << " " << (bar + "coming from foo") << std::endl;
```
Both have a potential memory leak. Number 1 is perfectly valid code that compiles without warning. Number 2 fails on line 3 with an error (`bar.size()`) and, with recent compilers, a warning (`bar + "coming from foo"` doesn't do what you think).

mexanick replied:
Ok, I rest my case that it would be better (more descriptive) to use an `std::string` return type here, but that's minor.

mexanick commented on gemonlinedb/src/common/ConfigurationManager.cc:
```diff
@@ -0,0 +1,167 @@
+                XercesDOMParser parser;
+                parser.setErrorHandler(errorHandler.get());
+                parser.setExternalSchemaLocation(
+                    ("https://github.com/cms-gem-daq-project/cmsgemos/tree/master/gemonlinedb/xml/schema/system-topology.xsd "
```
Should be a `const std::string` defined somewhere in a central location rather than hardcoded here (e.g. we can move the repo to gitlab etc.). Also a note: such code requires external `https` access which is at least cumbersome at P5 and might be not recommended way to go...

lmoureaux replied:
This syntax defines the location of the XML schema identified by `https://github.com/cms-gem-daq-project/...`. The second argument (after the whitespace) is its path in the local filesystem. No remote access is required.

The schema URL has to be written in XML files (when they are namespace-aware like this one), the `xsd` file and the loading code. It is hence part of the data format.

In the event of a migration to Gitlab, I'm hoping that a read-only archive would be kept for historical purpose.

mexanick replied:
Again, I would define it as a string constant in a well-known place, rather than burried somewhere in the code, so in case we need to change it, it would be more straightforward

lmoureaux replied:
Then give me a `configure` step to sync it with the `xsd` and example `xml`… I can do that with any modern build system, but not the xDAQ Makefiles.

Note that changing this value would result in old files not working anymore (that's what I meant with "part of the data format").

jsturdy replied:
Jumping in late here, I wouldn't set the schema URL to the VC file, rather to a site, e.g., where we'll host, e.g., the documentation and `yum` repos (yet to be fully configured, almost certainly somewhere in https://www.cern.ch/cmsgemdaq/sw/)
Then it can be "versioned" in the event the schema itself needs to be updated.
Additionally, as it's currently just a namespace identifier, I think we can leave it as is for now (but @lmoureaux can you add a `FIXME host for schema` comment next to it so my routine searches for `FIXME` will ensure this doesn't get lost?)

jsturdy commented:
My personal view regarding `auto`/`decltype` can be summarized as follows:
I don't care at all when they show up in the definition/implementation.
I've been reluctant to (over)use them in headers for the primary reason that explicit typing is useful as a documentation tool.

I don't want to be too much of a curmudgeonly old man, so I'll just say this:
* If it doesn't obfuscate usage, go for it
* If it aids the readability, go for it
* If it doesn't clearly satisfy the aforementioned points, ask if/why you think it's necessary

jsturdy commented on gemonlinedb/src/common/ConfigurationManager.cc:
```diff
@@ -0,0 +1,167 @@
+            try {
+                detail::xercesExceptionsToXcept([&]{
+                    // Deduce the search path for XML files
+                    std::string path = "/etc/opt/cmsgemos"; // FHS 3.0 standard location
```
`/etc/opt/cmsgemos.d` is created by the RPM build process, with an optional `/etc/opt/cmsgemos.conf` as some sort of package configuration probably
I might prefer `/opt/cmsgemos/etc/...` as a location (everything can then be referenced from `CMSGEMOS_ROOT`), but in the end, if it's packaged correctly, it won't matter

lmoureaux replied:
Referring to [section 3.7.4](https://refspecs.linuxfoundation.org/FHS_3.0/fhs/ch03s07.html#etcoptConfigurationFilesForOpt) of the Filesystem Hierarchy Standard, the path should be `/etc/opt/...`. The reasoning behind the standard seems to be that the complete configuration from a given host resides in one directory that can be backed up independently, or copied over to a clean install.

Files in the config path are host-specific in nature (and optional: XML will not be used at P5) so the packages do not include them. The intent is that sysadmins will maintain a set of files matching their setup in `/etc/opt/cmsgemos`. When users need to override (some of) them, they can set the environment variable with the same syntax as a traditional `PATH`.

`.d` directories usually have a special meaning: configuration fragments found in these directories will be merged together to produce the final config. `gemonlinedb` doesn't support this method of merging. Since `/etc/opt/cmsgemos.d` appears not to be used at all, I would  replace it by `/etc/opt/cmsgemos`. Most directories in `/etc` do not use the `.d` suffix anyway.

jsturdy replied:
The close linking between the build process for `cmsgemos` and `xdaq` would encourage adoption of a similar structure
```
[]% ll /opt/xdaq/
total 40
drwxr-xr-x.  2 root root  4096 Feb  1  2018 bin
drwxr-xr-x.  2 root root  4096 Feb  1  2018 config
drwxr-xr-x.  3 root root  4096 Feb  1  2018 etc
drwxr-xr-x. 50 root root  4096 Feb 14 15:13 htdocs
drwxr-xr-x. 59 root root  4096 Feb  1  2018 include
drwxr-xr-x.  4 root root 12288 Feb 22 12:15 lib
drwxr-xr-x.  2 root root  4096 Dec 16  2016 sbin
drwxr-xr-x.  4 root root  4096 Oct 25  2017 share
[]% ll /opt/xdaq/etc
total 48
-rw-r-xr-x. 1 root root  1826 Nov 14  2017 base.profile
drwxr-xr-x. 5 root root  4096 Dec 19  2016 CAENUpgrader
-rw-r-xr-x. 1 root root  3909 Nov 14  2017 default.profile
-rw-r-xr-x. 1 root root  1899 Nov 14  2017 default.profile.macosx
-rw-r-xr-x. 1 root root   462 Dec 16  2016 htaccess
-rw-r-xr-x. 1 root root    53 Dec 16  2016 htgroup
-rw-r-xr-x. 1 root root    19 Dec 16  2016 htpasswd
-rw-r-xr-x. 1 root root   255 Dec 16  2016 log4cplus.properties
-rwxr-xr-x. 1 root root 10765 Dec 16  2016 tinyproxy.conf
-rw-r-xr-x. 1 root root   435 Dec 16  2016 udpServer.sh
[]% ll /opt/cactus
total 16
drwxr-xr-x. 4 root root 4096 Aug  1  2018 bin
drwxr-xr-x. 4 root root 4096 Aug  1  2018 etc
drwxr-xr-x. 4 root root 4096 Aug  1  2018 include
drwxr-xr-x. 3 root root 4096 Aug  3  2018 lib
[]% ll /opt/cactus/etc
total 8
drwxr-xr-x. 2 root root 4096 Jan 31  2018 amc13
drwxr-xr-x. 4 root root 4096 Aug  1  2018 uhal
[]% ll /etc/opt
total 8
drwxr-x---. 2 gempro gempro 4096 Feb 14 15:55 cmsgemos.d
drwxr-xr-x. 9 root   root   4096 Mar 24  2018 rh
```

Though, on the other hand, the `xdaq` zone configurations are installed in
```
/etc/xdaq.d
```

So, given that
```
/opt/cmsgemos/etc
```
and 
```
/etc/opt/cmsgemos.d
```
are already created (and at least one of them used), unless there is a reasonable concern that you're attempting to resolve, one of the two should be used.

lmoureaux committed:
Use /opt/cmsgemos/etc instead of /etc/opt/cmsgemos

Apparently it is better to follow xDAQ and use a non-standard location.

Requested by @jsturdy on #266.

jsturdy reviewed (commented):
Last thing, there is a mix of `doxygen` syntax styles here, that we should normalize
I think I've used `@tag` everywhere in this repo (others of our repos have adopted `\tag`), and in this PR you have a mix of both.

lmoureaux committed:
Fix style of Doxygen comments

See PR #266

lmoureaux commented:
Fixed.

jsturdy reviewed (commented):
I lied, one more :-)
The member values prefixed by `///<` don't require any further tags, I think `doxygen` gets them right automatically
Then can you squash all of these commits (related to `doxygen` syntax)

jsturdy commented:
I meant these should be squashed into a single commit (though I should probably wait for the result of your "to be squashed" comment :-) )
```
* 41e43e7        (louis/feature/onlinedb-manager) More dox fixes to be squashed
* 1742871        ConfigurationManager.cc: \brief -> @brief
* 8cf4260        ConfigurationManager.h: \c -> @c
* bacb362        ConfigurationManager.h: \brief -> @brief
```

lmoureaux committed:
Add FIXME for Jared

See PR #266

lmoureaux commented:
Doxygen commits squashed.
